api/paidAction/index.js

```diff
@@ -0,0 +1,309 @@
+      result: await action.perform?.({ invoiceId: invoice.id, ...args }, context),
+      paymentMethod: 'OPTIMISTIC'
+    }
+  }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
+}
+
+async function performPessimisticAction (actionType, args, context) {
+  const { models, lnd } = context
+  const action = paidActions[actionType]
+
+  if (!action.supportsPessimism) {
+    throw new Error(`This action ${actionType} does not support pessimistic invoicing`)
+  }
+
+  if (context.hmac) {
+    return await models.$transaction(async tx => {
+      context.tx = tx
+
+      // make sure the invoice is HELD
+      const invoice = await verifyPayment(context)
+      args.invoiceId = invoice.id
+
+      await settleHodlInvoice({ secret: invoice.preimage, lnd })
+
+      return {
+        result: await action.perform(args, context),
+        paymentMethod: 'PESSIMISTIC'
+      }
+    }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
+  } else {
+    // just create the invoice and complete action when it's paid
+    return {
+      invoice: await createDbInvoice(actionType, args, context),
+      paymentMethod: 'PESSIMISTIC'
+    }
+  }
+}
+
+export async function retryPaidAction (actionType, args, context) {
+  const { models, me } = context
+  const { invoiceId } = args
+
+  const action = paidActions[actionType]
+  if (!action) {
+    throw new Error(`retryPaidAction - invalid action type ${actionType}`)
+  }
+
+  if (!me) {
+    throw new Error(`retryPaidAction - must be logged in ${actionType}`)
+  }
+
+  if (!action.supportsOptimism) {
+    throw new Error(`retryPaidAction - action does not support optimism ${actionType}`)
+  }
+
+  if (!action.retry) {
+    throw new Error(`retryPaidAction - action does not support retrying ${actionType}`)
+  }
+
+  if (!invoiceId) {
+    throw new Error(`retryPaidAction - missing invoiceId ${actionType}`)
+  }
+
+  context.user = await models.user.findUnique({ where: { id: me.id } })
+  return await models.$transaction(async tx => {
+    context.tx = tx
+    context.optimistic = true
+
+    // update the old invoice to RETRYING, so that it's not confused with FAILED
+    const { msatsRequested, actionId } = await tx.invoice.update({
+      where: {
+        id: invoiceId,
+        actionState: 'FAILED'
+      },
+      data: {
+        actionState: 'RETRYING'
+      }
+    })
+
+    context.cost = BigInt(msatsRequested)
+    context.actionId = actionId
+
+    // create a new invoice
+    const invoice = await createDbInvoice(actionType, args, context)
+
+    return {
+      result: await action.retry({ invoiceId, newInvoiceId: invoice.id }, context),
+      invoice,
+      paymentMethod: 'OPTIMISTIC'
+    }
+  }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
+}
+
+// TODO: switch these back before going to production
+const OPTIMISTIC_INVOICE_EXPIRE = { minutes: 10 } // { hours: 1 }
+const PESSIMISTIC_INVOICE_EXPIRE = { minutes: 10 } // { minutes: 10 }
```

Comment:
Actually, I think we want 10 minutes for these. We aren't able to determine if they're failed until they expire, so anything longer could lead to items or zaps being much later the person intended. At least when they're short, we can prompt them to retry in a reasonable time frame.